Sending/Displaying messages fix

tianhaoyao opened the pull request:
Messages are now correctly displayed (in order, correct recipient). Security issue could be not including a sender with each message, so I refactored the code to always include sender. This information could be very useful when refactoring sockets, so that the sockets will limit emitting information to the recipient and sender only.

Before: 
const reqBody = {
      text: event.target.text.value,
      recipientId: this.props.otherUser.id,
      conversationId: this.props.conversationId,
      sender: this.props.conversationId ? null : this.props.user,
    };

Now:
const reqBody = {
      text: event.target.text.value,
      recipientId: this.props.otherUser.id,
      conversationId: this.props.conversationId,
      exists: this.props.conversationId ? true : false,
      sender: this.props.user,
    };

tmattmac commented on client/src/components/ActiveChat/Messages.js:
```diff
@@ -1,14 +1,24 @@
-    <Box>
-      {messages.map((message) => {
+    <Box className={classes.chatDisplay}>
+      {messages.sort((a,b) => (a.id > b.id) ? 1 : ((b.id > a.id) ? -1 : 0)).map((message) => {
```
Sorting here means it'd have to sort each time the component re-renders. You should be able to fix the order of the messages on the backend when they first get requested.

Also, it'd be preferred to sort by the created date of the message rather than the ID.

tmattmac commented on server/bin/www:
```diff
@@ -43,6 +43,7 @@ io.on("connection", (socket) => {
   socket.on("new-message", (data) => {
     socket.broadcast.emit("new-message", {
       message: data.message,
+      recipientId: data.recipientId,
       sender: data.sender,
     });
```
This still sends the socket message to every connected socket, so it wouldn't deal with the socket issue. But you don't have to worry about the sockets until issue 4.

tmattmac commented on client/src/store/utils/reducerFunctions.js:
```diff
@@ -1,7 +1,7 @@
-  if (sender !== null) {
+  const { message, sender, exists } = payload;
+  // if not exists, then make a new convo
+  if (exists !== true && exists !== 1) {
```
Is there a particular case where `exists` would be 1?

tianhaoyao replied:
I am not sure why, but the exists property gets translated into a 1 from true. I printed from both sides (sending and receiving), the sender's exists is true and the receiver is 1. I've just decided to keep everything 1 or 0 instead of true or false now.

tmattmac replied:
It turns out the default socket.io parser apparently doesn't support booleans. 1 and 0 are truthy and falsy, respectively, so they should still work in any conditions you provide them to. Here, specifically, you could just use `!exists` which will treat it as a boolean.

tmattmac reviewed (commented):
The security issue we mentioned is related to the API (not sockets) allowing any user to post a message to a conversation regardless of whether they're involved in the conversation. We'd like to see you tackle this issue, as well as address the other comments I've left.

tmattmac commented on server/bin/www:
```diff
@@ -33,15 +34,27 @@ const io = require("socket.io")(server);
     if (!onlineUsers.includes(id)) {
       onlineUsers.push(id);
+      userMap[id] = socket.id
+      console.log(userMap)
```
```suggestion

```

tmattmac commented on server/bin/www:
```diff
@@ -33,15 +34,27 @@ const io = require("socket.io")(server);
 
   socket.on("new-message", (data) => {
-    socket.broadcast.emit("new-message", {
+    console.log(data)
```
```suggestion

```

tmattmac commented on server/bin/www:
```diff
@@ -52,6 +65,7 @@ io.on("connection", (socket) => {
     if (onlineUsers.includes(id)) {
       userIndex = onlineUsers.indexOf(id);
       onlineUsers.splice(userIndex, 1);
+      delete userMap[id];
```
Couldn't you just use the `userMap` as the list of online users instead?

tmattmac commented on server/routes/api/conversations.js:
```diff
@@ -20,9 +20,9 @@ router.get("/", async (req, res, next) => {
       attributes: ["id"],
-      order: [[Message, "createdAt", "DESC"]],
+      order: [[Message, "createdAt", "ASC"]],
       include: [
-        { model: Message, order: ["createdAt", "DESC"] },
+        { model: Message, order: ["createdAt", "ASC"] },
```
Make sure that this still shows conversations in order of latest first.

tmattmac commented on server/bin/www:
```diff
@@ -33,15 +34,27 @@ const io = require("socket.io")(server);
+    const senderId = data.sender.id;
+    const recipientId = data.recipientId;
+
+    socket.to(userMap[senderId]).broadcast.emit("new-message", {
+      message: data.message,
+      recipientId: data.recipientId,
+      sender: data.sender,
+    });
```
The user sending the message shouldn't need to receive the message; they can just update the local state with the message they're sending.

tmattmac commented on server/bin/www:
```diff
@@ -33,15 +34,27 @@ const io = require("socket.io")(server);
+      recipientId: data.recipientId,
+      sender: data.sender,
+    });
+    socket.to(userMap[recipientId]).broadcast.emit("new-message", {
```
If you're using `to`, `broadcast` is unnecessary.

tmattmac commented on server/routes/api/messages.js:
```diff
@@ -1,43 +1,68 @@
-      senderId,
-      recipientId
-    );
+    const token = req.cookies["JWT"];
+    if (token) {
+      jwt.verify(token, process.env.SESSION_SECRET, (err, decoded) => {
+        if (err) {
+          return next();
+        }
+        User.findOne({
+          where: { id: decoded.id },
+        }).then(async (user) => {
+          if (user.dataValues.id !== sender.id) {
+            res.sendStatus(403);
```
Doesn't the middleware take care of this already?

tmattmac commented on server/routes/api/messages.js:
```diff
@@ -1,43 +1,68 @@
+          if (user.dataValues.id !== sender.id) {
+            res.sendStatus(403);
+          } else {
+            // if we already know conversation id, we can save time and just add it to message and return
+            if (conversationId) {
+              const message = await Message.create({
+                senderId,
+                text,
+                conversationId,
+              });
+              return res.json({ message, sender });
+            }
```
The user is still able to send messages to a conversation they aren't involved in. You will need to check the conversation ID to see if the user is in the conversation before sending the message.

tianhaoyao commented:
I addressed everything @tmattmac mentioned above. I have also implemented a read/unread feature using sockets, the backend, and redux. I know the feature is not fully complete yet (need a count for unread messages saved in redux, and UI changes reflecting on that) but it is still a step closer. The next step is to save an unread messages count in each conversation in redux.

tmattmac commented on client/src/components/ActiveChat/OtherUserBubble.js:
```diff
@@ -1,48 +1,72 @@
-        <Typography className={classes.usernameDate}>
-          {otherUser.username} {time}
-        </Typography>
+        {props.read ? (
+          <Typography className={classes.usernameDate}>
+            {otherUser.username} {time} read
+          </Typography>
+        ) : (
+          <Typography className={classes.usernameDate}>
+            {otherUser.username} {time} unread
+          </Typography>
+        )}
```
Since this ternary operator only affects the read/unread text, you can just use it for the text rather than the whole component.

tmattmac commented on client/src/components/ActiveChat/SenderBubble.js:
```diff
@@ -1,41 +1,52 @@
       <Box className={classes.bubble}>
         <Typography className={classes.text}>{text}</Typography>
       </Box>
+      {display ? (
+        <Avatar
+          alt={otherUser.username}
+          src={otherUser.photoUrl}
+          className={classes.avatar}
+        />
+      ) : null}
```
```suggestion
      {display && (
        <Avatar
          alt={otherUser.username}
          src={otherUser.photoUrl}
          className={classes.avatar}
        />
      )}
```

tmattmac commented on client/src/components/Sidebar/Sidebar.js:
```diff
@@ -29,7 +29,7 @@ const Sidebar = (props) => {
       <Typography className={classes.title}>Chats</Typography>
       <Search handleChange={handleChange} />
-      {conversations
+      {conversations.slice(0).reverse()
```
You shouldn't have to reverse the conversations on the client side; you can retrieve the messages on the server in the correct order once.

tianhaoyao replied:
I'm trying to figure out the correct order for displaying these by changing the order attributes in `Conversation.findAll`, but I'm having no luck. It seems that the outer order `order: [[Message, "createdAt", "DESC"]],` must be kept as `DESC` to be able to change the order of the conversations at all. However, doing so means I cannot change the order of the inner `{ model: Message, order: [Message, "createdAt", "DESC"] }` as it does not seem to have any effect whatever I change this to. The result of this is the correct conversation order but the messages are upside down for each chat (old messages at the bottom, new at the top). 

I'm stuck on this because if I change the outer order to `order: [[Message, "createdAt", "ASC"]],`, the conversation order is not preserved.

```js 
...
order: [[Message, "createdAt", "DESC"]],
      include: [
        { model: Message, order: ["createdAt", "DESC"] }, //this order has no effect
        ...
```

tianhaoyao replied:
I've decided to go with this solution for now, please let me know if there is a way to do it directly in the findAll.

```js
for (let i = 0; i < conversations.length; i++) {
      const convo = conversations[i];
      const convoJSON = convo.toJSON();
      convoJSON.messages = convoJSON.messages.reverse();
      ...
```

tmattmac replied:
For efficiency, it'd be better to get the messages in the correct order and then sort the conversations by the last message's `createdAt` timestamp.

tmattmac commented on server/routes/api/conversations.js:
```diff
@@ -61,6 +61,25 @@ router.get("/", async (req, res, next) => {
+          convoJSON.messages[i].senderId !== convoJSON.otherUser.id &&
+          convoJSON.messages[i].read
+        ) {
+          theirUneadIndex = i;
```
This seems to be a typo.

tmattmac commented on server/routes/api/read.js:
```diff
@@ -0,0 +1,46 @@
+    const senderVerify = req.user.id;
+    const { msgId, senderId, recipientId, read } = req.body;
+    const token = req.cookies["JWT"];
+    if (token) {
+      jwt.verify(token, process.env.SESSION_SECRET, (err, decoded) => {
+        if (err) {
+          return next();
+        }
+        User.findOne({
+          where: { id: decoded.id },
+        }).then(async (user) => {
```
This process should be taken care of by the middleware. The logged in user is already accessible as `req.user`.

tmattmac commented on server/routes/api/read.js:
```diff
@@ -0,0 +1,46 @@
+          if (user.dataValues.id !== senderVerify) {
+            res.sendStatus(403);
+          } else {
+            Message.findOne({
+              where: {id: msgId}
+            }).then(async (message) => {
+              //save to database
+              message.read = read;
+              await message.save();
```
You can update the message with a single DB query instead of two separate ones.

tmattmac commented on server/routes/api/conversations.js:
```diff
@@ -61,6 +61,25 @@ router.get("/", async (req, res, next) => {
+      let myUnreadCount = 0;
+      let theirReadIndex = 0;
+
+      for (let i = 0; i < convoJSON.messages.length; i++) {
```
It'd be more efficient to calculate this from the end of the array.

tmattmac commented on client/src/store/index.js:
```diff
@@ -32,4 +32,34 @@ const store = createStore(rootReducer, /* preloadedState, */ composeEnhancers(
 
+export const updateUnreadCount = (state, msgId) => {
+  let conversationKey = -1
+  for(let j = 0; j < state.length; j++) {
+    for(let i = 0; i < state[j].messages.length; i++){
+      if(state[j].messages[i].id === msgId){
+        conversationKey = j
+        break;
+        
+      }
+    }
+  }
```
Is there any way you could make this more efficient, perhaps by sending the conversation ID along with the message ID?

tmattmac reviewed (commented):
In the future, make sure you're making changes for different issues in different branches and making a new PR accordingly.
